### ISSUE TITLE
Improvements to indirect_adapter

### ISSUE THREAD

**Issue by Morwenn** [Morwenn/cpp-sort#166]
`indirect_adapter` is based on the first algorithm that I found at the time that did the job properly, but it hasn't been improved as much as it could be:
- It only supports random-access iterators.
- It stores iterators while pointers are surely enough for some scenarios.
- There are other algorithms to investigate.

Most notably the following algorithms are worth having a look at:
- (BSL) Boost.Algorithm has [`apply_permutation`][apply-permutation] which can be used to reorder the elements of a collection according to a collection of positions, apparently based on a [series of blog posts][raymond-chen] by Raymond Chen.
- (zlib) [plf_indiesort][indiesort] is a sorting algorithm wrapper that takes a given sorting algorithm and uses it to sort by pointer. It works with forward iterators if needed.
- (BSL) Boost.Sort has a bunch of indirect sorts, I'm not sure what they do exactly but they might be of interest.

A roadmap of actions to take to give some love to `indirect_adapter`:
- [x] Analyze the different solutions.
- [x] Implement different flavours of `indirect_adapter` based on the different algorithms.
- [x] Benchmark (the indiesort benchmarks can probably give meaningful results).
- [ ] Steal what is worth stealing (licenses seem compatible).
- [x] Add tests for forward and bidirectional iterators.
- [x] Update documentation.

  [apply-permutation]: https://www.boost.org/doc/libs/1_73_0/libs/algorithm/doc/html/algorithm/Misc.html#the_boost_algorithm_library.Misc.misc_inner_algorithms.apply_permutation
  [indiesort]: https://github.com/mattreecebentley/plf_indiesort
  [raymond-chen]: https://devblogs.microsoft.com/oldnewthing/20170105-00/?p=95125

**Comment by Morwenn** [comment]
Sorting an `std::deque<std::array<int, 100>>` of 10^6 elements. Only the first element of each array is compared every time to benchmark the principal use case of `indirect_adapter`: sorting a collection of expensive to copy elements with a cheap comparison.

![Indirect adapters benchmark bar plot](https://i.imgur.com/LcZ5LUl.png)

As can be seen the 1.7.0 implementation of `indirect_adapter` based on [mountain sort](https://github.com/Morwenn/mountain-sort) performs better than the new experimental one based on [indiesort](https://github.com/mattreecebentley/plf_indiesort). That said the results are close enough, and indiesort can sort collections that expose forward and bidirectional iterators, making it a worthwhile addition to the library for that specific use case. Mountain sort is kept for the implementation targeting random-access iterators.

**Comment by Morwenn** [comment]
Somehow I didn't manage to get Boost.Algorithm's [`apply_permutation`](https://www.boost.org/doc/libs/1_73_0/libs/algorithm/doc/html/algorithm/Misc.html#the_boost_algorithm_library.Misc.misc_inner_algorithms.apply_permutation) to work in `indiect_adapter`, but it still provided a small optimization that I I used to improve the indiesort guts.

The optimization in question happens when putting an element in its place in the post-processing step: instead of indicating the elements already sorted with a `sorted` sentinel of type `difference_type`, we assign the sorted's elements `original_index` the index of its own position. The main post-processing loop only considers elements that are not in their final place, so we take advantage of that to simplify the condition and get rid of the `sorted` sentinel.

Another small advantage we get from this change: we now support exotic iterator types for whom `difference_type` does not support the `std::numeric_limit` template (for example `__uint128_t` without standard library support).

**Comment by Morwenn** [comment]
I reproduced the previous benchmark with mountain sort vs. Boost.Sort's [`indirect_sort`](https://github.com/boostorg/sort/blob/develop/include/boost/sort/common/indirect.hpp#L132-L144):

![Indirect adapters benchmark bar plot](https://i.imgur.com/M9JrBBv.png)

It performs slightly better for the best case, but is otherwise slower than mountain sort, and seems comparatively slower than indiesort too. I won't try to use this one in cpp-sort.

**Comment by Morwenn** [comment]
I tried to rebase the algorithm on `apply_permutation` once again: the basic version was slower than the one based on mountain sort, with pretty much the same speed as the one based on Boost.Sort's `indirect_sort`. I tried to change it to avoid `swap` once again to avoid performing extra relocations, but I didn't get a working version, and it looks enough like other algorithms to consider that it won't perform visibly better. I'm closing this issue, enough work has been done.
